File: internal/daemon/backends/git.go
package backends

import (
	"fmt"
	"log/slog"
	"strings"

	"github.com/mschulkind-oss/swarf/internal/gitexec"
	"github.com/mschulkind-oss/swarf/internal/paths"
)

type GitBackend struct{}

func (g *GitBackend) Sync(storePath string) SyncResult {
	slog.Info("sync: staging all changes", "store", storePath)
	gitexec.AddAll(storePath)

	status := gitexec.StatusPorcelain(storePath)
	if strings.TrimSpace(status) == "" {
		slog.Info("sync: no changes to commit")
		return SyncResult{Success: true, Message: "No changes to sync", FilesChanged: 0}
	}

	nFiles := 0
	for _, line := range strings.Split(strings.TrimSpace(status), "\n") {
		if strings.TrimSpace(line) != "" {
			slog.Info("sync: staged", "file", strings.TrimSpace(line))
			nFiles++
		}
	}

	s := "s"
	if nFiles == 1 {
		s = ""
	}
	msg := fmt.Sprintf("auto: sync %d file%s", nFiles, s)
	slog.Info("sync: committing", "message", msg)
	if err := gitexec.Commit(storePath, msg); err != nil {
		slog.Error("sync: commit failed", "path", storePath, "err", err)
		return SyncResult{Success: false, Message: "Commit failed: " + err.Error(), FilesChanged: 0}
	}
	stampNow(paths.LastCommitFile)
	slog.Info("sync: committed locally", "files", nFiles)

	remote := gitexec.RemoteURL(storePath, "")
	if remote != "" {
		slog.Info("sync: pushing to remote", "remote", remote)
		if err := gitexec.Push(storePath, "origin"); err != nil {
			slog.Warn("sync: push failed (will retry later)", "remote", remote, "err", err)
			return SyncResult{Success: true, Message: fmt.Sprintf("Committed %d files locally, push failed", nFiles), FilesChanged: nFiles}
		}
		stampNow(paths.LastPushFile)
		slog.Info("sync: pushed to remote successfully", "remote", remote, "files", nFiles)
	} else {
		slog.Info("sync: no git remote configured, skipping push")
	}

	return SyncResult{Success: true, Message: fmt.Sprintf("Synced %d files", nFiles), FilesChanged: nFiles}
}

func (g *GitBackend) HasChanges(storePath string) bool {
	return strings.TrimSpace(gitexec.StatusPorcelain(storePath)) != ""
}

File: src/swarf/daemon/backends/git.py
"""Git sync backend for the swarf daemon."""

from __future__ import annotations

import logging
from pathlib import Path

from swarf.daemon.backends import SyncBackend, SyncResult
from swarf.git import git_add_all, git_commit, git_push, git_remote_url, git_status_porcelain

logger = logging.getLogger(__name__)


class GitBackend(SyncBackend):
    """Sync via git add/commit/push."""

    def sync(self, swarf_path: Path) -> SyncResult:
        """Add, commit, and push changes in the swarf directory."""
        git_add_all(swarf_path)

        status = git_status_porcelain(swarf_path)
        if not status.strip():
            return SyncResult(success=True, message="No changes to sync", files_changed=0)

        n_files = len([line for line in status.strip().splitlines() if line.strip()])

        try:
            git_commit(swarf_path, f"auto: sync {n_files} file{'s' if n_files != 1 else ''}")
        except Exception:
            logger.exception("Failed to commit in %s", swarf_path)
            return SyncResult(success=False, message="Commit failed", files_changed=0)

        # Push only if a remote is configured
        if git_remote_url(swarf_path):
            try:
                git_push(swarf_path)
            except Exception:
                logger.warning("Push failed for %s (will retry later)", swarf_path)

        return SyncResult(success=True, message=f"Synced {n_files} files", files_changed=n_files)

    def has_changes(self, swarf_path: Path) -> bool:
        """Check for uncommitted changes."""
        status = git_status_porcelain(swarf_path)
        return bool(status.strip())
